src/Analyzer.Cli.UnitTests/CommandLineParserTests.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.Azure.Templates.Analyzer.Cli;
+using System.Linq;
+using System;
+using System.IO;
+
+namespace Analyzer.Cli.FunctionalTests
+{
+    [TestClass]
+    public class CommandLineParserTests
+    {
+        private CommandLineParser _commandLineParser;
+
+        [TestInitialize]
+        public void TestInit()
+        {
+            _commandLineParser = new CommandLineParser();
+        }
+
+        [DataTestMethod]
+        [DataRow("Path does not exist", 2, DisplayName = "Invalid template file path provided")]
+        [DataRow("\\CommandLineParserTests.cs", 4, DisplayName = "Path exists, not an ARM template.")]
+        [DataRow("\\CommandLineParserTests.cs", 3, "--report-format", "Sarif", DisplayName = "Path exists, Report-format flag set, --output-file-path flag not included.")]
+        [DataRow("\\CommandLineParserTests.cs", 1, "--parameters-file-path", DisplayName = "Path exists, Parameters-file-path flag included, but no value provided.")]
+        [DataRow("\\AppServicesLogs-Failures.json", 5, DisplayName = "Issues found in the template")]
+        [DataRow("\\AppServicesLogs-Passes.json", 0, DisplayName = "Success")]
+        public void AnalyzeTemplate_ValidInputValues_ReturnExitCode(string relativeTemplatePath, int expectedExitCode, params string[] additionalParams)
+        {
+            var args = new string[] { "analyze-template" , String.Concat(Directory.GetCurrentDirectory(), relativeTemplatePath)}; 
+            args = args.Concat(additionalParams).ToArray();
+            var result = _commandLineParser.InvokeCommandLineAPIAsync(args);
+
+            Assert.AreEqual(expectedExitCode, result.Result);
+        }
+
+        [DataTestMethod]
+        [DataRow("\\Configuration.json", 1, DisplayName = "Provided parameters file is not a parameters file")]
+        [DataRow("\\Parameters.json", 5, DisplayName = "Provided parameters file correct, issues in template")]
+        public void AnalyzeTemplate_ParameterFileParamUsed_ReturnExitCode(string relativeParametersFilePath, int expectedExitCode)
+        {
+            var templatePath = String.Concat(Directory.GetCurrentDirectory(), "\\AppServicesLogs-Failures.json");
+            var parametersFilePath = String.Concat(Directory.GetCurrentDirectory(), relativeParametersFilePath);
+            var args = new string[] { "analyze-template", templatePath, "--parameters-file-path", parametersFilePath };
+            var result = _commandLineParser.InvokeCommandLineAPIAsync(args);
+
+            Assert.AreEqual(expectedExitCode, result.Result);
+        }
+
+        [TestMethod]
+        public void AnalyzeTemplate_ConfigurationFileParamUsed_ReturnExitCode()
```

Comment:
I was not sure if I should bring in Analyzer.Reports.ConsoleReportWriter to be able to read the test outputs (ie. would benefit the ConfigurationFile test). These are functional tests so it should test everything end-to-end, but at the same time, those are already tested in different projects and I did not want to add an additional project dependency to this project. Thoughts?

Reply:
I think if the sarif output is tested somewhere else it should be OK to not test here. If we want to test that something was actually output to the correct path, we can probably test that part of the functionally if it's not tested in the Reports tests without importing the writer. i.e. If you hard-code the expected sarif output somewhere in the test project and just string compare it to whatever is written to the outputfile. Only thin about this is that if we make updates to the sarif output properties, then this test will have to be updated, but that shouldn't be an issue or a big deal.

Reply:
Just want to point out that functional tests don't necessarily have to be end-to-end.  Functional will test a larger area than a unit, but it can still be smaller than end to end.  Just depends on what the test goal is.

Reply:
Also @anaismiller you can run the raw CLI and get the output text from the full run with `System.Diagnostics.Process.Start()`, so you don't have to worry about the ReportWriter classes.

Reply:
Thank you. Yane3628's point about hard-coded expectation and updating the test if a property changes and JohnathonMohr;s point that functional tests != end-to-end made me decide to leave the test as it is.